File: backend/app/schemas/notification.py
"""
Notification Schemas
Pydantic v2 models for notification management
"""

from datetime import datetime
from typing import Optional, Dict, Any, List

from pydantic import BaseModel, Field, ConfigDict, field_validator

from app.models.notification import NotificationType


class NotificationBase(BaseModel):
    """Base notification schema"""
    title: str = Field(..., max_length=200, description="Notification title")
    message: str = Field(..., description="Notification message")
    notification_type: NotificationType = Field(
        default=NotificationType.INFO,
        description="Notification type"
    )
    action_url: Optional[str] = Field(None, max_length=500, description="Optional action URL")
    action_label: Optional[str] = Field(None, max_length=100, description="Optional action button label")
    # Note: Field name is 'metadata' for API
    # The model uses 'notification_metadata' attribute (with 'metadata' property for Pydantic)
    metadata: Optional[Dict[str, Any]] = Field(None, description="Additional metadata (JSON)")


class NotificationCreate(NotificationBase):
    """Notification creation schema"""
    user_id: int = Field(..., description="User ID to send notification to")
    
    @field_validator('title')
    @classmethod
    def validate_title(cls, v: str) -> str:
        """Validate title"""
        if not v or not v.strip():
            raise ValueError('Title cannot be empty')
        return v.strip()
    
    @field_validator('message')
    @classmethod
    def validate_message(cls, v: str) -> str:
        """Validate message"""
        if not v or not v.strip():
            raise ValueError('Message cannot be empty')
        return v.strip()


class NotificationUpdate(BaseModel):
    """Notification update schema"""
    read: Optional[bool] = Field(None, description="Read status")
    action_url: Optional[str] = Field(None, max_length=500, description="Action URL")
    action_label: Optional[str] = Field(None, max_length=100, description="Action label")
    # Note: Field name is 'metadata' for API
    # The model uses 'notification_metadata' attribute (with 'metadata' property for Pydantic)
    metadata: Optional[Dict[str, Any]] = Field(None, description="Metadata")


class NotificationResponse(NotificationBase):
    """Notification response schema"""
    id: int
    user_id: int
    read: bool
    read_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True  # Allow both field name and alias
    )


class NotificationListResponse(BaseModel):
    """Notification list response schema"""
    notifications: List[NotificationResponse]
    total: int
    unread_count: int
    skip: int = 0
    limit: int = 100


class NotificationUnreadCountResponse(BaseModel):
    """Unread count response schema"""
    unread_count: int
    user_id: int

File: apps/web/src/types/notification.ts
/**
 * Notification Types
 * TypeScript types aligned with backend Pydantic schemas
 */

export type NotificationType = 'info' | 'success' | 'warning' | 'error';

export interface NotificationBase {
  title: string;
  message: string;
  notification_type: NotificationType;
  action_url?: string | null;
  action_label?: string | null;
  metadata?: Record<string, unknown> | null;
}

export interface NotificationCreate extends NotificationBase {
  user_id: number;
}

export interface NotificationUpdate {
  read?: boolean;
  action_url?: string | null;
  action_label?: string | null;
  metadata?: Record<string, unknown> | null;
}

export interface Notification extends NotificationBase {
  id: number;
  user_id: number;
  read: boolean;
  read_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface NotificationListResponse {
  notifications: Notification[];
  total: number;
  unread_count: number;
  skip: number;
  limit: number;
}

export interface NotificationUnreadCountResponse {
  unread_count: number;
  user_id: number;
}

// Extended notification for UI components (includes UI-specific fields)
export interface NotificationUI extends Notification {
  // UI-specific fields (not from backend)
  icon?: React.ReactNode;
  avatar?: string;
  sender?: {
    name: string;
    avatar?: string;
  };
}

// Helper type for notification filters
export interface NotificationFilters {
  read?: boolean;
  notification_type?: NotificationType;
  skip?: number;
  limit?: number;
}

